src/block_tanh.rs

```diff
@@ -0,0 +1,166 @@
+use libm::tanhf;
+use std::any::Any;
+use std::error::Error;
+use crate::block_helpers;
+use crate::feature_buffer;
+use crate::graph;
+use crate::graph::BlockGraph;
+use crate::model_instance;
+use crate::port_buffer;
+use crate::regressor;
+use regressor::BlockTrait;
+
+pub struct BlockTanh {
+    pub num_inputs: usize,
+    pub input_offset: usize,
+    pub output_offset: usize,
+}
+
+pub fn new_tanh_block(
+    bg: &mut graph::BlockGraph,
+    input: graph::BlockPtrOutput,
+) -> Result<graph::BlockPtrOutput, Box<dyn Error>> {
+    let num_inputs = bg.get_num_output_values(vec![&input]);
+    assert!(num_inputs != 0);
+    let block = Box::new(BlockTanh {
+        output_offset: usize::MAX,
+        input_offset: usize::MAX,
+        num_inputs: num_inputs,
+    });
+    let mut block_outputs = bg.add_node(block, vec![input])?;
+    assert_eq!(block_outputs.len(), 1);
+    Ok(block_outputs.pop().unwrap())
+}
+
+impl BlockTrait for BlockTanh {
+    fn as_any(&mut self) -> &mut dyn Any {
+        self
+    }
+
+    fn allocate_and_init_weights(&mut self, _mi: &model_instance::ModelInstance) {}
+
+    fn get_num_output_slots(&self) -> usize {
+        1
+    }
+
+    fn get_num_output_values(&self, output: graph::OutputSlot) -> usize {
+        assert!(output.get_output_index() == 0);
+        return self.num_inputs;
+    }
+
+    fn set_input_offset(&mut self, input: graph::InputSlot, offset: usize) {
+        assert!(input.get_input_index() == 0);
+        self.input_offset = offset;
+    }
+
+    fn set_output_offset(&mut self, output: graph::OutputSlot, offset: usize) {
+        assert!(output.get_output_index() == 0);
+        self.output_offset = offset;
+    }
+
+    #[inline(always)]
+    fn forward_backward(
+        &mut self,
+        further_blocks: &mut [Box<dyn BlockTrait>],
+        fb: &feature_buffer::FeatureBuffer,
+        pb: &mut port_buffer::PortBuffer,
+        update: bool,
+    ) {
+        debug_assert!(self.output_offset != usize::MAX);
+        debug_assert!(self.input_offset != usize::MAX);
+        debug_assert!(self.num_inputs > 0);
+
+        unsafe {
+            for i in 0..self.num_inputs as usize {
+                let x = *pb.tape.get_unchecked_mut(self.input_offset + i);
+
+                // for now using libm tanh computation. once we establish a baseline,
```

Comment:
This is a good idea - having full dependency just for computing tanh seems like a lot? (might have missed other uses)

Reply:
Ok. In case we'd want to roll our own, [This one ](https://varietyofsound.wordpress.com/2011/02/14/efficient-tanh-computation-using-lamberts-continued-fraction/)should do a pretty decent trick perhaps
```
float fast_tanh(float x){
  float x2 = x * x;
  float a = x * (135135.0f + x2 * (17325.0f + x2 * (378.0f + x2)));
  float b = 135135.0f + x2 * (62370.0f + x2 * (3150.0f + x2 * 28.0f));
  return a / b;
}

```

Reply:
cool! I hope I will get to test that as well.
I speculate @andraztori's approach is to replace those with lookup tables (which don't involve any computation...), but I think we'll get there only after we see the value - so I will proceed testing